ddtrace/contrib/botocore/services/sqs.py

```diff
@@ -0,0 +1,279 @@
+            trace_data[PROPAGATION_KEY_BASE_64] = get_pathway(pin, endpoint_service, dsm_identifier, span)
+        inject_trace_data_to_message_attributes(trace_data, entry, endpoint_service)
+
+
+def inject_trace_to_sqs_or_sns_message(params, span, endpoint_service=None, pin=None, data_streams_enabled=False):
+    # type: (Any, Span, Optional[str], Optional[Pin], Optional[bool]) -> None
+    """
+    :params: contains the params for the current botocore action
+    :span: the span which provides the trace context to be propagated
+    :endpoint_service: endpoint of message, "sqs" or "sns"
+    :pin: patch info for the botocore client
+    :data_streams_enabled: boolean for whether data streams monitoring is enabled
+
+    Inject trace headers and DSM info into MessageAttributes for the SQS or SNS record
+    """
+    trace_data = {}
+    HTTPPropagator.inject(span.context, trace_data)
+
+    if data_streams_enabled:
+        dsm_identifier = None
+        if endpoint_service == "sqs":
+            dsm_identifier = get_queue_name(params)
+        elif endpoint_service == "sns":
+            dsm_identifier = get_topic_arn(params)
+
+        trace_data[PROPAGATION_KEY_BASE_64] = get_pathway(pin, endpoint_service, dsm_identifier, span)
+
+    inject_trace_data_to_message_attributes(trace_data, params, endpoint_service)
+
+
+def patched_sqs_api_call(original_func, instance, args, kwargs, function_vars):
+    params = function_vars.get("params")
+    trace_operation = function_vars.get("trace_operation")
+    pin = function_vars.get("pin")
+    endpoint_name = function_vars.get("endpoint_name")
+    operation = function_vars.get("operation")
+
+    message_received = False
+    func_run = False
+    func_run_err = None
+    ctx = None
+    start_ns = None
+    result = None
+
+    if config.botocore.propagation_enabled:
+        if endpoint_name == "sqs" and operation == "ReceiveMessage":
+            # Ensure we have Datadog MessageAttribute enabled
+            if "MessageAttributeNames" not in params:
+                params.update({"MessageAttributeNames": ["_datadog"]})
+            elif "_datadog" not in params["MessageAttributeNames"]:
+                params.update({"MessageAttributeNames": list(params["MessageAttributeNames"]) + ["_datadog"]})
+
+            # Ensure we have AWS attribute enabl
+            # ed for interoperability with dd-trace-java aws propagation
+            if "AttributeNames" not in params:
+                params.update({"AttributeNames": ["AWSTraceHeader"]})
+            elif "AWSTraceHeader" not in params["AttributeNames"]:
+                params.update({"AttributeNames": list(params["AttributeNames"]) + ["AWSTraceHeader"]})
+
+            try:
+                start_ns = time_ns()
+                func_run = True
+                # run the function before in order to extract possible parent context before starting span
+                result = original_func(*args, **kwargs)
```

Comment:
The original func call is not inside the tracer.trace(). This seems to be a variation from previous implementation? Or did I misunderstand this?

Reply:
Yes, this was done as a workaround to limitations with how we start spans in Python and was previously discussed with the Python guild. Basically, we need the parent context before starting a span, since there's no functionality to re-parent a span once its created. Only way to do this is by running the original function, then starting the span (with a parent context if one is extracted)

Reply:
Oh I missed the "ReceiveMessage" part.  This makes sense.